clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +371,311 @@ impl FunctionType {
+            TypeSignature::TupleType(expected_tuple_type),
+        ) => {
+            for (name, atom_field_type) in atom_tuple_type.get_type_map() {
+                match expected_tuple_type.field_type(name) {
+                    Some(expected_field_type) => {
+                        inner_type_check_type(
+                            db,
+                            contract_context,
+                            atom_field_type,
+                            expected_field_type,
+                            clarity_version,
+                            depth + 1,
+                            tracker,
+                        )?;
+                    }
+                    None => {
+                        return Err(CheckErrors::TypeError(
+                            expected_type.clone(),
+                            atom_field_type.clone(),
+                        )
+                        .into())
                     }
                 }
             }
         }
-        Ok(returns.clone())
+        (
+            TypeSignature::CallableType(CallableSubtype::Trait(atom_trait_id)),
+            TypeSignature::CallableType(CallableSubtype::Trait(expected_trait_id)),
+        ) => {
+            if atom_trait_id != expected_trait_id {
+                let atom_trait = lookup_trait(
+                    db,
+                    contract_context,
+                    &atom_trait_id,
+                    clarity_version,
+                    tracker,
+                )?;
+                let expected_trait = lookup_trait(
+                    db,
+                    contract_context,
+                    expected_trait_id,
+                    clarity_version,
+                    tracker,
+                )?;
+                trait_check_trait_compliance(
+                    db,
+                    contract_context,
+                    &atom_trait_id,
+                    &atom_trait,
+                    expected_trait_id,
+                    &expected_trait,
+                    clarity_version,
+                    tracker,
+                )?;
+            }
+        }
+        (
+            TypeSignature::CallableType(CallableSubtype::Principal(contract_identifier)),
+            TypeSignature::CallableType(CallableSubtype::Trait(expected_trait_id)),
+        ) => {
+            let contract_to_check = db
+                .load_contract(&contract_identifier)
+                .ok_or(CheckErrors::NoSuchContract(contract_identifier.to_string()))?;
+            runtime_cost(ClarityCostFunction::AnalysisFetchContractEntry, tracker, 1)?;
+            let expected_trait = lookup_trait(
+                db,
+                contract_context,
+                expected_trait_id,
+                clarity_version,
+                tracker,
+            )?;
+            contract_to_check.check_trait_compliance(expected_trait_id, &expected_trait)?;
+        }
+        (
+            TypeSignature::ListUnionType(types),
```

Comment:
Hey, can you remind me what `TypeSignature::ListUnionType` represents?  Is it just a type signature for any sequence, or does it also include optional and response types?

Reply:
Sorry, probably poor naming choice. It's meant to hold a set of unresolved but potentially compatible callable types, used to determine the least super type for a list of callables. For example, different principal literals and trait instances might all be able to match a common trait type.

Maybe it would be more clear if I instead moved this inside of the `CallableSubtype` enum and called it `CallableUnion` or something?

Reply:
On second thought, I don't love that idea, because a value of this type is not callable, it contains callables.